uwblueprint/starter-code-v2
# Integrate TS FileStorageService into EntityService

## Notion ticket link

Closes #80 



## Implementation description
* Updated models to have a file name field
* Updated entity request DTO to have a filePath field and response DTO to have a fileName field
* Updated EntityService for Postgres and mongo to call FileStorageService methods
* Updated entity routes , added separate get endpoint for retrieving file by fileUUID
* Added file type validation
* Updated create/update entity GraphQL mutations to take a FileUpload type parameter (documentation https://www.apollographql.com/docs/apollo-server/data/file-uploads/)
* Added separate GraphQL query to retrieve file by fileUUID
* Refactor FileStorageService (changing the bucket field to bucketName field to prevent firebase functions from being called before initializeApp())



## Steps to test
REST:
(using Postman)
1. Set the body to form-data and add the body field (as stringified JSON) and file field (postman lets you upload files).
<img width="936" alt="image" src="https://user-images.githubusercontent.com/37782734/125882389-3af38b05-e17f-4c55-951b-87e344a8c4b2.png">

```
{
    "stringField": "test 1",
    "intField": 1,
    "enumField": "A",
    "stringArrayField": ["test1"],
    "boolField": true
}
```
(if you want an example)
2. Send the POST request to entities endpoint to create an entity. Verify in firebase console that the file is created.
3. Try creating another entity whose mime type does not belong to text/plain, application/pdf, image/png, image/jpeg, image/gif. Verify that there is a validation error.
4. Try creating another entity without file parameter. Verify that this works (file field is optional).
5. Send a GET request to /entities endpoint. Verify that the file name is returned.
6. Send a GET request to /entities/files/{fileUUID}. Verify that you get the signed URL.
7. Send a PUT request to this entity's endpoint, with a different file. Verify that the file is updated.
8. Send a PUT request to this entity's endpoint, with no file field. Verify that the file is deleted in firebase console.
9. Send a PUT request to this entity's endpoint, with a file field. Verify that a new file UUID is generated and that it appears in firebase console.
10. Send a DELETE request to this entity's endpoint. Verify that the file is deleted in firebase console.
11. Repeat with the other EntityService (mongo or Postgres)

GraphQL:
1. Download [Altair](https://altair.sirmuel.design) (or use postman with graphql setting, haven't tested it with this though)
2. Add the file at the bottom of the window (click variables)
<img width="611" alt="image" src="https://user-images.githubusercontent.com/37782734/125883199-c0ee9422-1182-4b3d-9a64-cf6dbbb1ae3f.png">
You can use this mutation to create entity.

```
mutation($file: Upload) {
  createEntity(
    entity: {
      stringField: "hi", 
      intField: 2,
      stringArrayField: ["hi"],
      boolField: true,
      enumField: A
    }, file: $file) {
    id
    stringField
    intField
    boolField
    enumField
    stringArrayField
    fileName
  }
}
```
3. You can use this query to get a file signed url.
```
query {
   file(fileUUID: "cat.png")
}
```
4. Repeat similar steps as REST for update entity, you can use this mutation.
```
mutation($file: Upload) {
  updateEntity(
    id: 1,
    entity: {
      stringField: "hi", 
      intField: 2,
      stringArrayField: ["hi"],
      boolField: true,
      enumField: A
    }, file: $file) {
    id
    stringField
    intField
    boolField
    enumField
    stringArrayField
    fileName
  }
}
```
5. Repeat similar step for deleting entity, you can use this mutation
```
mutation {
  deleteEntity(id: 1)
}
```
6. Repeat with the other EntityService (mongo or Postgres)


## What should reviewers focus on?
* Update functionality is complex - does everything look ok?


## Checklist
